tck/src/main/java/org/eclipse/microprofile/lra/tck/TckRecoveryTests.java

```diff
@@ -0,0 +1,167 @@
+    private URL deploymentURL;
+    
+    private Client deployementClient;
+    private WebTarget deployementTarget;
+
+    @Rule
+    public TestName testName = new TestName();
+        
+    @Before
+    public void before() throws URISyntaxException {
+        LOGGER.info("Running test: " + testName.getMethodName());
+        
+        deployementClient = ClientBuilder.newClient();
+        deployementTarget = deployementClient.target(deploymentURL.toURI());
+    }
+    
+    @After
+    public void after() {
+        deployementClient.close();
+    }
+
+    @Deployment(name = DEPLOYMENT_NAME, managed = false)
+    public static WebArchive deploy() {
+        return TckTestBase.deploy(DEPLOYMENT_NAME);
+    }
+    
+    @Test
+    public void testOneFailureWithCancelAfterRecovery() {
+        // deploy the test service
+        deployer.deploy(DEPLOYMENT_NAME);
+        
+        // invoke phase 1 inside the container (start new LRA and enlist resource)
+        Response response1 = deployementTarget.path(RecoveryResource.RECOVERY_RESOURCE_PATH)
+            .path(RecoveryResource.PHASE_1).request().get();
+        
+        Assert.assertEquals(200, response1.getStatus());
+        URI lra = URI.create(response1.readEntity(String.class));
+
+        // kill the test service while LRA is still active
+        deployer.undeploy(DEPLOYMENT_NAME);
+
+        // start the test service again
+        deployer.deploy(DEPLOYMENT_NAME);
+
+        // invoke phase 2 inside the container (cancel the LRA and verify that the
+        // @Compensate method was called on the enlisted resource)
+        Response response2 = deployementTarget.path(RecoveryResource.RECOVERY_RESOURCE_PATH)
+            .path(RecoveryResource.PHASE_2)
+            .request()
+                .header(RecoveryResource.LRA_HEADER, lra)
+                .get();
+        
+        Assert.assertEquals(response2.readEntity(String.class), 200, response2.getStatus());
+        
+        deployer.undeploy(DEPLOYMENT_NAME);
+    }
+
+    @Test
+    public void testOneFailureWithCancelDuringRecovery() {
+        // deploy the test service
+        deployer.deploy(DEPLOYMENT_NAME);
+
+        // invoke phase 1 inside the container (start new LRA and enlist resource) which cancels after 500 millis
+        Response response1 = deployementTarget.path(RecoveryResource.RECOVERY_RESOURCE_PATH)
+            .path(RecoveryResource.PHASE_1)
+            .queryParam("timeout", true)
+            .request().get();
+
+        Assert.assertEquals(200, response1.getStatus());
+        URI lra = URI.create(response1.readEntity(String.class));
+
+        // kill the test service while LRA is still active
+        deployer.undeploy(DEPLOYMENT_NAME);
+        
+        // wait for the timeout cancellation of the LRA while the service is still down
+        // Compensate should be attempted to be called while the participant service is down
+        try {
+            Thread.sleep(500);
```

Comment:
I think that makes sense as when the participant or the service starting the LRA specifies timeout then this is a declaration that it will keep the resources for compensation for this period (or slightly more). If we allow Compensation for a timeout in the eventual point in the future after the timelimit is reached this information may no longer be present or relevant and thus we may get into invalid states.

Reply:
Can't we say something like when the period has elapsed the LRA should make reasonable effort to only transition to cancelled rather than mandate a specific action at a specific time? I am very hesitant to describe something as being done at an exact time. Even then, perhaps the LRA checks if it is not timed out, this check passes, then it moves to complete the LRA but due to a thread scheduling delay the actual complete call takes place _after_ the timeout has occured. For example:
Time 1: lra.setTimeout(time3)
Time 2: if (!lraHasTimedOut(lra.getTimeout))
Time 3:  // server overload, delay of an undefined period of time before next instruction, this LRA is now in theory timed out
Time 4: lra.complete() // Impossible to prevent

Reply:
There should not be any suggestion that a participant can ever decided ot perform it's compensation procedure unilaterally and still expect an overall atomic outcome for the transaction to be achieved.

A resource should guarantee to keep the resources available for compensation until the LRA directs it otherwise.

Reply:
>when the period has elapsed the LRA should make reasonable effort to only transition to cancelled rather than mandate a specific action at a specific time? 

Yes, this is what we currently have in the spec (or how I interpret it, we have no specific time requirements)
But this TCK test mandate that we have specific time requirements. During undeploy, wait and deploy, it assumes the timeout is completely processed.

We have to accept that when we say 'eventual consistency' and 'at a certain point in the future', etc ..., you basically loose the ability to really test this.

We have already introduced some 'tricks' to trigger the recovery (to make things testable) but maybe we should go the opposite direction and accept that parts of the spec can't be tested.

Reply:
I understand your points but if you look at it all from the user perspective what does timeout mean for a participant? My interpretation was that I as a participant can, for instance, say that I will be able to compensate for the next hour after then I'll lose the ability to undo the performed action (e.g. the bought ticket cannot be reimbursed anymore). How would we deal with such situation if actual Compensate calls would be received in eventual manner after Cancellation?

Reply:
It is also worthwhile noting that the time out is a hard requirement in the sense that if the timeout is reached then the only possible outcome is for the LRA to eventually cancel. That gives users of the spec some transactional guarantees and makes the feature very useful.

The `tricks` we use are fully configurable by the implementation. Any reasonable implementation should be able to tweak the recovery delay such that it is fully within spec.

Although in a real world scenario the participants may never fully converge (the management environment may never restart a crashed participant for example) this should never happen in a controlled TCK testing scenario, ie this behaviour should be fully testable by the TCK.

Reply:
> There should not be any suggestion that a participant can ever decided ot perform it's compensation procedure unilaterally and still expect an overall atomic outcome for the transaction to be achieved.
> 
That is correct and we have the FailedToComplete/Compensate states in the model to cover this.
> A resource should guarantee to keep the resources available for compensation until the LRA directs it otherwise.
Or it should remember that it was unable to compensate until it is explicitly told to forget.

But on the question of timers, they don't have to cancel the LRA as soon as they fire but the implementation must ensure that the the only possible outcome for the LRA is `Cancelled` or `FailedToCancel`.

Reply:
Isn't this scenario covered by the `FailedToComplete` status. The spec allows failure states. I mean that it is a hard requirement in the sense that when the implementation sees that the timer has expired it is free to invoke compensation activities at its leisure provided it eventually tells them all to compensate.

With respect to your scenario, we have a state model and the request to complete and the expiration of a timer both affect the state. It is a race condition as to which decision is made durable first (and the loosing update will fail since it would be requesting an invalid state transition). A distributed implementation needs to carefully orchestrate the various parties to ensure that conflicts do not arise (it is okay for participants to enter failure states but the implementation should not be asking different participants to perform different actions since that would increase failure rates). An implementation that uses a centralised coordinator can avoid this situation without serious difficulties. A distributed implementation would need to perform transactional updates to a shared store (a database for example) to avoid conflicting state transitions.

Reply:
>we have a state model and the request to complete and the expiration of a timer both affect the state. It is a race condition as to which decision is made durable first (and the loosing update will fail since it would be requesting an invalid state transition). A distributed implementation needs to carefully orchestrate the various parties to ensure that conflicts do not arise (it is okay for participants to enter failure states but the implementation should not be asking different participants to perform different actions since that would increase failure rates). An implementation that uses a centralised coordinator can avoid this situation without serious difficulties. A distributed implementation would need to perform transactional updates to a shared store (a database for example) to avoid conflicting state transitions.

This very useful and required info to put in the spec.

Reply:
so we should document that in the timeout spec paragraph or javadoc.

On the issue at hand, would it be possible to wait for the short or long delay after the timeout should be triggered to make sure that each implementation has enough time to actually cancel the LRA? Otherwise, as it was mentioned by Rudy, this scenario is not testable.

Reply:
I guess a short delay is probably enough although a possible implementation could be to mark the LRA as `Cancelling` when the timer expires and it then cancels the LRA on its next `scheduled recovery pass` which could be a while. Maybe the safest option is to "trigger a recovery pass".

Reply:
trigger recovery pass will be a hard to implement in this case since we control when the container is running the test is not running inside of the container so for trigger recovery we would need:

- kill the container
- start a new container with only the ability to trigger recovery
- trigger recovery in the new container
- kill the new container
- start the original container

Can we agree whether this is acceptable or we should just include the short or long wait here, please?

@mmusgrov @rdebusscher

Reply:
I possibly found a workaround but it depends on the LraRecoveryService to be created by the implementation (now its defined as CDI bean) -- it can be both, but if it depends on CDI injection that wouldn't work in the tests running outside of the container nevertheless. It works in Narayana, @rdebusscher would it be possible for you also, please?

Reply:
Let's keep it simple then and just used one of the existing delays (no delay or applyShortConsistencyDelay() would both work for the narayana implementation)

Reply:
ok

Reply:
Not convinced that adding just a delay can solve the issue. Since there are no timing requirements when the @Compensate should be called after transaction is cancelled due to timeout (as we discussed, there are no timing requirements in general) we cannot guarantee that all the 'events' occur within the time frame we expect them to happen for this test.
The external triggering by LraRecoveryService could solve it but adding it as required component within the implementation is not an option.

Reply:
@rdebusscher 
1/ this is a misunderstanding - my workaround still covers only TCK -- poor choice of words on my side. I implemented it with Arquillian resource injection thus the reason why CDI would not work if we would include it.

2/ I was thinking about this timeout discussion. How the cancel from timeout differs from a normal user indicated cancel? Both cases mean eventual compensations so we already have this issue solved with short / long delays in TCK I think. So calling short delay here is the same as for instance in `TckTests#joinWithOneResource` (random pick).

Reply:
So this is the idea of the test

- LRA started
- LRA times out / Participant goes down
- @Compensate callback (to Participant who is down)
- Participant started again
- @Compensate callback again (retry)

But how can we enforce that the @Compensate callback is called during the Participant is down when there are no timing requirements in the spec about when the @Compensate should be called.

>If execution is to run with an active LRA and the associated class contains other methods
annotated with `@Compensate` and `@Complete` then these methods will be associated with
the active LRA. They will be invoked when the LRA is subsequently cancelled or closed.

Could be changed to

>If execution is to run with an active LRA and the associated class contains other methods
annotated with `@Compensate` and `@Complete` then these methods will be associated with
the active LRA. They **MUST be invoked immediately** when the LRA is subsequently cancelled or closed.


And yes, I realised now that we have an issue in general, so also the test `TckTests#joinWithOneResource` assumes it is immediately (which is not defined in the spec) and what does `immediately` even mean?
And the wait we use there is just a technical one so that required acrtions can be performed and not spec related.

Reply:
My understanding is that short delay is there for eventual compensations. So TCK implementations can say what time is required for them to deliver the eventual call to `@Compensate` methods. Or in other words, by setting short delay (implementors do that through MP config) you are saying that you guarantee **in the TCK** that if cancel/close is called then after a short delay the compensate calls have been invoked (doesn't mean successfully but the implementation must have tried before the short delay is elapsed). 

So I think we don't need to change the spec text. This is really just a workaround for TCK so we are able to test eventual consistency. Maybe we can revisit the short/long delay solution before GA if you think it's necessary, but this PR should be good to go?

Reply:
If we want to be sure that the implementation tries to call the Participant when it is undeployed, we should use the LraRecoveryService. Otherwise this test will be flaky.

```
deployer.undeploy();
lraRecoveryService.triggerRecovery(lraId);
deployer.deploy();
```

Reply:
that would mean (and was my original intention) to rewrite all short/long delays to use triggerRecovery(). 

Can we make assumption on LraRecoveryService to be instantiable without CDI?

Reply:
@rdebusscher It seems like you are hijacking this PR when the [issue has already been discussed](https://github.com/eclipse/microprofile-lra/pull/240#discussion_r335336214) (and from my point of view resolved). Do we need to move this conversion back to one of the issues?

Reply:
@mmusgrov I'm not hijacking this PR. I'm concerned that this specific test is very flaky due to the timing requirements.

And what stops you of merging this and just carry on on your own?

Reply:
@xstefank 

>Can we make assumption on LraRecoveryService to be instantiable without CDI?

Yes

Reply:
Ok. So I can rewrite it to use LraRecoveryService. Can we agree on that this is the right way to go for this PR? @mmusgrov @rdebusscher

Reply:
LraRecoveryService is in my opinion the only way to make sure we can bypass some timing issues and thus making the tests stable.
I though LraRecoveryService was already agreed on and the way to remove all those waits so that Tests could be executed faster.

Reply:
> @mmusgrov I'm not hijacking this PR. I'm concerned that this specific test is very flaky due to the timing requirements.

In the [thread where I justified why the TCK can test eventual consistency](https://github.com/eclipse/microprofile-lra/pull/240#discussion_r335336214) there were no further comments so my assumption was that my arguments were valid. But in your later comments on the PR you seem to be still arguing that we cannot test eventual consistency in the TCK.

> And what stops you of merging this and just carry on on your own if you are convinced that everyone else is wrong and you are correct?

Again I have had no followups to my justification of why we can have TCK tests for eventual consistency. Who else is in agreement with Rudy?

Reply:
We must make some assumption if we want to finish TCK in some sane time :)

I personally think that with LraRecoveryService we give the implementation the control of the replaying and delivering participant callback invocations. So if the implementation still wants to use timeouts it will still be possible but not mandated. It can also make the TCK runs potentially faster as we are not going to wait even when the calls have been already delivered.

However, I agree that maybe for normal cancel/close (or timeout) this may be a little unnecessary -- this means short delay. A short delay is there only for potential small delays that can happen between cancel call an Compensate invocations (similarly for close). But this makes an assumption that implementation will call Compensate right when it receives cancel call which as Rudy mentioned is not strictly required by the specification. So LraRecoveryService may have little overhead but it is the correct way of handling this case also.

So I am +1 on at least try to rewrite TCK to utilize LraRecoveryService. I will have time for this next week and I would like to first verify with Narayana that it is actually possible and how it affects TCK performance.

Reply:
The short and long delays are arbitrary but under the control of the implementation that is under test. For implementations that do things immediately for a particular test the short delay can be set to zero. For tests where there is a need for some recovery operations then the long delay kind of helps but the better solution is to have an explicit `recover` operation which we had and which has now been reintroduced. That is a TCK only feature but it means that an implementation is able to pass the TCK even under the assumptions of `eventual consistency`. [If, in the future, there is an implementation that cannot be validated in this way then we would need to rethink the TCK].

@xstefank Why do you need to rewrite the TCK. All that should be required is to replace calls to long delay with calls to LraRecoveryService#retry

Reply:
We can for sure replace only long delays and that would be sufficient as you are right that Narayana indeed has the short delay set to 0. But @rdebusscher is right that eventual consistency means that once it can be 0ms other time 2s and another time 3 years so I understand his concern that for some implementations this actually can be performance degradation. This is a trade-off we need to decide. @rdebusscher I would like to hear your opinion first.

Reply:
Are you talking about the short delays. If so then yes it would be good to get rid of that too. But not by triggering recovery since the Narayana implementation does not need that. So something like:

```java
    boolean waitForCallbacks(); // returns after an attempt to notify all participants was made
    boolean retryCallbacks(); // some callbacks previously failed so request redelivery
```

Reply:
Ok, since we diverged a lot from this PR intentions :) I've created new issues to track this for individual long and short delays:

- https://github.com/eclipse/microprofile-lra/issues/249
- https://github.com/eclipse/microprofile-lra/issues/250

 @rdebusscher @mmusgrov can we merge this PR as is (with a short delay for timeout) and adjust it later in these two issues, please?

Reply:
I though you had agreed to replace the `Thread.sleep(500);` call with `applyShortConsistencyDelay`

Will you also include a description in those two issues you raised of the rationale of why we need them and why the existing approach of using delays is unsatisfactory. Without that context an external contributor would not be able to contribute.

Reply:
@xstefank @mmusgrov the way we are discussion things at the moment isn't working at all. People aren't replying to the correct answer, you can't indicate to what question you reply, Github is not even showing all comments anymore,  etc...

This way, it is complete waste of my and everyone else his time.

And yes, this comment on the PR isn't either a good place for the above comment but that is the way we seem to work ....

Reply:
If you want to reply to a specific comment github has a "Quote Reply" option in the ... pull down.

Reply:
@rdebusscher Part of the problem is that we are fundamentally at odds on how this spec should work which generates lots of discussion. Are the issues a better way to discuss things? If so then as soon as it becomes clear on a PR that there are disagreements we should move back onto the issue for further discussion. Or do you have another suggestion?

Reply:
> Can we make assumption on LraRecoveryService to be instantiable without CDI?

@xstefank I think it is okay for the TCK but I would prefer to avoid reliance on CDI in the spec itself since that constrains which frameworks/containers we can integrate with.

Reply:
@rdebusscher let's take the discussion on how to better discuss issues on gitter please. I agree that a lot of at least my PRs often diverge into some different issues.

@mmusgrov 
> > Can we make assumption on LraRecoveryService to be instantiable without CDI?
> 
> @xstefank I think it is okay for the TCK but I would prefer to avoid reliance on CDI in the spec itself since that constrains which frameworks/containers we can integrate with.

This is still only TCK. Basically the original idea was to have it instantianable through service loader but I changed to only be a CDI bean. Now we need to go a step back to something similar as service loader. I think that you actually agree with the original statement (Can we make assumption on LraRecoveryService to be instantiable without CDI?)